src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/HashPartitioningQueryAndScanMapper.java

```diff
@@ -0,0 +1,108 @@
+        this.physicalTable = physicalTable;
+        this.keyMapper = keyMapper;
+        this.maxBucketNumber = keyMapper.getNumberOfBucketsPerVirtualTable() - 1;
+    }
+
+    @Override
+    ScanResult executeScanInternal(AmazonDynamoDB amazonDynamoDB, ScanRequest scanRequest) {
+        RequestIndex requestIndex = getRequestIndexAndMapIndexName(new ScanRequestWrapper(scanRequest, null, null));
+        PrimaryKey virtualPk = requestIndex.getVirtualPk();
+        PrimaryKey physicalPk = requestIndex.getPhysicalPk();
+
+        QueryRequest queryRequest = new QueryRequest()
+            .withTableName(physicalTable.getTableName())
+            .withIndexName(scanRequest.getIndexName())                  // already mapped index name
+            .withExclusiveStartKey(scanRequest.getExclusiveStartKey())  // map later for easy access of virtual HK value
+            .withFilterExpression(scanRequest.getFilterExpression())
+            .withProjectionExpression(scanRequest.getProjectionExpression())
+            .withExpressionAttributeNames(scanRequest.getExpressionAttributeNames())
+            .withExpressionAttributeValues(scanRequest.getExpressionAttributeValues())
+            .withReturnConsumedCapacity(scanRequest.getReturnConsumedCapacity())
+            .withConsistentRead(scanRequest.getConsistentRead())
+            .withLimit(scanRequest.getLimit());
+        QueryRequestWrapper queryRequestWrapper = new QueryRequestWrapper(queryRequest, null, null);
+
+        // see which bucket we should start at. if there's an ExclusiveStartKey, then use the corresponding bucket.
+        // otherwise, start from the beginning at 0.
+        int bucket = 0;
+        if (scanRequest.getExclusiveStartKey() != null) {
+            AttributeValue virtualHkValue = scanRequest.getExclusiveStartKey().get(virtualPk.getHashKey());
+            bucket = keyMapper.getBucketNumber(virtualPk.getHashKeyType(), virtualHkValue);
+            // map to physical start key
+            applyExclusiveStartKey(queryRequestWrapper, requestIndex);
+        }
+        // build initial physical query request, where the key condition specifies the physical HK that corresponds to
+        // the initial bucket.
+        String physicalHkFieldPlaceholder = MappingUtils.getNextFieldPlaceholder(queryRequestWrapper);
+        String physicalHkValuePlaceholder = MappingUtils.getNextValuePlaceholder(queryRequestWrapper);
+        queryRequest.withKeyConditionExpression(physicalHkFieldPlaceholder + " = " + physicalHkValuePlaceholder)
+            .addExpressionAttributeNamesEntry(physicalHkFieldPlaceholder, physicalPk.getHashKey())
+            .addExpressionAttributeValuesEntry(physicalHkValuePlaceholder, keyMapper.toPhysicalHashKey(bucket));
+
+        // keep moving forward pages until we find at least one record or reach end
+        QueryResult queryResult;
+        while ((queryResult = amazonDynamoDB.query(queryRequest)).getItems().isEmpty()) {
```

Comment:
So we query one bucket and pass the user-specified limit down?

The doc says: `A single Scan operation reads up to the maximum number of items set (if using the Limit parameter) or a maximum of 1 MB of data and then apply any filtering to the results using FilterExpression.`

It seems like ideally, we'd implement that contract (for random partitioning as well), but we probably don't have enough data returned from the `ScanResult` to make that distinction, right? (We are running into the same problem with streams.)

Reply:
Yeah... I don't think there's anything in ScanResult / QueryResult telling us that. We could estimate the size of each returned item and aggregate the sizes, but I don't know how inaccurate that would be.